### PR TITLE
v.next interface for package:file/file.dart

### PR DESCRIPTION
**Pull request by tvolkert**
This changes the interface to chiefly expose the FileSystem
abstraction and deal in native dart:io types where at all possible:

- Expose all dart:io file-system static methods as instance methods
  on `FileSystem`, to allow injectable implementations (e.g. in-memory,
  local, mock, etc.)
- Create FileSystemEntity, File, Directory, Link as abstract classes
  that implement their native counterparts and add any extra methods
  we choose in this library (initially, only `get fileSystem => Filesystem`)

By going this route, it implies that each FileSystem implementation will
need to provide types that implement both the sync and async APIs (since
they implement the interfaces in native dart:io). However, I still plan
to provide SynchronousFileSystem & AsynchronousFileSystem, which will simply
throw UnsupportedError for the APIs they don't support.

This change will allow existing libraries/apps to seamlessly start using
FileSystem as a drop-in replacement for existing direct dart:io usage.

### CODE REVIEW

**Comment by tvolkert** [comment]
@matanlurey @cbracken

**Comment by tvolkert** [comment]
Note: the Travis breakage is expected, as this was sent primarily to get an LGTM on the interface.  If it looks good, I'll merge despite Travis failures and send a follow-on restructuring to un-break things.

**Comment by tvolkert** [comment]
@cbracken they all implement FileSystemEntity, which adds a link back to the underlying FileSystem.

**Comment by cbracken** [comment]
Ah yup! lgtm from me.

lgtm

File: lib/src/interface/file_system.dart [hunk 1]
```diff
@@ -8,15 +12,98 @@ abstract class FileSystem {
...
+  /// Creates a directory object pointing to the current working directory.
+  Directory get currentDirectory;
+
+  /// Sets the current working directory to the specified path. The new value
+  /// set can be either a [Directory] or a [String].
+  ///
+  /// Relative paths will be resolved by the underlying file system
+  /// implementation (meaning it is up to the underlying implementation to
+  /// decide whether to support relative paths).
+  set currentDirectory(dynamic path);
```
Comment (matanlurey) [inline lib/src/interface/file_system.dart]: You can just omit dynamic, doesn't add any specific value here.

File: lib/src/interface/file_system.dart [hunk 2]
```diff
@@ -8,15 +12,98 @@ abstract class FileSystem {
...
+  /// `false`, and [path] points to a link
   ///
   /// If the [path] does not point to a file system object or an error occurs
   /// then [FileSystemEntityType.NOT_FOUND] is returned.
-  Future<FileSystemEntityType> type(String path, {bool followLinks: true});
+  io.FileSystemEntityType typeSync(String path, {bool followLinks: true});
+
+  /// Checks if [`type(path)`](type) returns [io.FileSystemEntityType.FILE].
+  Future<bool> isFile(String path) => type(path)
+      .then((io.FileSystemEntityType type) => type == io.FileSystemEntityType.FILE);
```
Comment (matanlurey) [inline lib/src/interface/file_system.dart]: A little cleaner:

```dart
Future<bool> isFile(String path) => await type(path) == io.FileSystemEntityType.FILE;
```

Comment (matanlurey) [inline lib/src/interface/file_system.dart]: (You can apply this throughout this file if you're able to)

**Comment by tvolkert** [comment]
All comments addressed - merging.

**Pull request merged**